cupy/core/core.pyx

```diff
@@ -861,34 +861,91 @@ cdef class ndarray:
             axis (int): Axis along which to sort. Default is -1, which means
                 sort along the last axis.
 
-        .. note::
-           For its implementation reason, :func:`cupy.ndarray.partition` fully
-           sorts the given array as :meth:`cupy.ndarray.sort` does. It also
-           does not support ``kind`` and ``order`` parameters that
-           :func:`numpy.partition` supports.
-
         .. seealso::
             :func:`cupy.partition` for full documentation,
             :meth:`numpy.ndarray.partition`
 
         """
-        ndim = self.ndim
+
+        if cupy.issubdtype(self.dtype, complex):
+            raise NotImplementedError('Sorting arrays with dtype \'{}\' is '
+                                      'not supported'.format(self.dtype))
+
+        cdef Py_ssize_t ndim = self.ndim
+
+        if ndim == 0:
+            raise ValueError('Sorting arrays with the rank of zero is not '
+                             'supported')
+
+        if not self._c_contiguous:
+            raise NotImplementedError('Sorting non-contiguous array is not '
+                                      'supported.')
+
         if axis < 0:
             axis += ndim
         if not (0 <= axis < ndim):
             raise _AxisError('Axis out of range')
 
+        if axis == ndim - 1:
+            data = self
+        else:
+            data = cupy.rollaxis(self, axis, ndim).copy()
+
         length = self.shape[axis]
         if isinstance(kth, int):
             kth = kth,
+        max_k = 0
         for k in kth:
             if k < 0:
                 k += length
             if not (0 <= k < length):
                 raise ValueError('kth(={}) out of bounds {}'.format(k, length))
+            if max_k < k:
+                max_k = k
+
+        # For simplicity, max_k is round up to the power of 2.
+        max_k = max(32, 1 << max_k.bit_length())
```

Comment:
If `max_k` is the power of 2, it is round up to the **next** power of 2. This behavior is expected, but it may be confusing. The argument `k` of `partition` specifies the index; on the other hand, the kernel collects the first `max_k` elements. I will update code and comment to clarify what we want to do.